docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+
+The builtin subcommands shall include network clients for interacting with local
+p2p nodes and the pub-sub bus. The vast majority of new functionality will,
+however, be built on top of the [`Storage`](#storage) primitive. If and when it
+becomes desirable to prototype such functionality in other programming languages
+(ie. without linking to the storage library directly) it will be considered to
+expose the storage interface as a network service itself.
+
+> Note that it does not make much sense to proxy networked subcommands when
+> binding to a socket. Frontends will need to connect to the respective services
+> directly.
+
+A comprehensive specification of the (initial) set of builtins is beyond the
+scope of this document.
+
+
+### Browser Applications
+
+In practice, browser applications are limited to HTTP or WebSockets for network
+calls, filesystem operations must be delegated to a backend service. This poses
+two security concerns:
+
+1. The backend must bind to a well-known TCP port, which is then accessible by
+   all other programs on the same host. Moreover, browsers can be tricked by
+   remote sites into issuing requests against the local host.
+2. Public-key based authentication APIs (such as [WebAuthn][webauthn]) are not
+   yet mature / flexible / portable enough to allow a browser frontend to prove
+   access to the (shared) private key to a backend process.
+
+For these reasons, we devise the following restrictions:
+
+* TCP-based backend services MUST NOT expose privileged operations.
+* Applications built using browser-based frameworks (such as [electron]) MUST
+  proxy service calls through the privileged `main` process, which is
+  responsible for acquiring signatures via the `ssh-agent` protocol where
+  necessary, and calling the backend service via [IPC](#ipc).
+
+Since preferences of web developers regarding the granularity of remote calls
+vary, they may choose to implement their own IPC backend, possibly re-composing
+the library functionality of [CLI](#cli) subcommands.
+
+Browser applications SHOULD be packaged in a way such that their main backend
+conforms to [Process Orchestration](#process-orchestration), and include other
+services via the package manager's dependency management mechanism. If that is
+not desired (e.g. [AppImage] bundles), they MUST sandbox all service
+dependencies, and MUST NOT interact with services installed outside the sandbox
+(ie. all configuration and state must be relative to the bundle root, process
+management is up to the application, and separate key material must be used).
+
+
+## Future Work
+
+The main omission of this RFC is to explore platform specifics for Windows
+targets. This is mainly because the author has stopped using Windows two decades
+ago, and was never very keen to understand the platform's idiosyncrasies. He
+conjectures, however, that equivalents to the system services proposed here
+exist, and would be grateful for pointers.
```

Comment:
The blessed way seems to be via PowerShell: https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_keymanagement

The strange thing about windows is that there are so many different ways to avoid having to use their platform idiosyncrasies, and instead just use UNIX tools: Cygwin, MSYS2, MinGW, WSL2, ...

I can’t even tell what the differences are, nor what developers typically use / have installed, or if that matters at all. Maybe these days PowerShell is so powerful that it’s kind of the emacs of windows and one can do anything we’d want on top of it?